3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/MailgunAdapter.scala

```diff
@@ -0,0 +1,201 @@
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics
+package snowplow
+package enrich
+package common
+package adapters
+package registry
+
+// Java
+import java.net.URI
+import org.apache.http.client.utils.URLEncodedUtils
+
+// Scala
+import scala.collection.JavaConversions._
+import scala.util.control.NonFatal
+import scala.util.{Try, Success => TS, Failure => TF}
+
+// Scalaz
+import scalaz._
+import Scalaz._
+
+// Jackson
+import com.fasterxml.jackson.core.JsonParseException
+
+// json4s
+import org.json4s._
+import org.json4s.JsonDSL._
+import org.json4s.jackson.JsonMethods._
+
+// Iglu
+import iglu.client.{
+  SchemaKey,
+  Resolver
+}
+
+// This project
+import loaders.CollectorPayload
+import utils.{JsonUtils => JU}
+
+/**
+ * Transforms a collector payload which conforms to
+ * a known version of the StatusGator Tracking webhook
+ * into raw events.
+ */
+object MailgunAdapter extends Adapter {
+
+  // Vendor name for Failure Message
+  private val VendorName = "Mailgun"
+
+  // Tracker version for an Mailgun Tracking webhook
+  private val TrackerVersion = "com.mailgun-v1"
+
+  // Expected content type for a request body
+  private val ContentTypes = List("application/x-www-form-urlencoded", "multipart/form-data")
+  private val ContentTypesStr = ContentTypes.mkString(" or ")
+
+  // Schemas for reverse-engineering a Snowplow unstructured event
+  private val EventSchemaMap = Map (
+    "bounced"      -> SchemaKey("com.mailgun", "message_bounced", "jsonschema", "1-0-0").toSchemaUri,
+    "clicked"      -> SchemaKey("com.mailgun", "message_clicked", "jsonschema", "1-0-0").toSchemaUri,
+    "complained"   -> SchemaKey("com.mailgun", "message_complained", "jsonschema", "1-0-0").toSchemaUri,
+    "delivered"    -> SchemaKey("com.mailgun", "message_delivered", "jsonschema", "1-0-0").toSchemaUri,
+    "dropped"      -> SchemaKey("com.mailgun", "message_dropped", "jsonschema", "1-0-0").toSchemaUri,
+    "opened"       -> SchemaKey("com.mailgun", "message_opened", "jsonschema", "1-0-0").toSchemaUri,
+    "unsubscribed" -> SchemaKey("com.mailgun", "recipient_unsubscribed", "jsonschema", "1-0-0").toSchemaUri
+  )
+
+  /**
+   * Converts a CollectorPayload instance into raw events.
+   *
+   * A Mailgun Tracking payload contains one single event
+   * in the body of the payload, stored within a HTTP encoded
+   * string.
+   *
+   * @param payload The CollectorPayload containing one or more
+   *        raw events as collected by a Snowplow collector
+   * @param resolver (implicit) The Iglu resolver used for
+   *        schema lookup and validation. Not used
+   * @return a Validation boxing either a NEL of RawEvents on
+   *         Success, or a NEL of Failure Strings
+   */
+def toRawEvents(payload: CollectorPayload)(implicit resolver: Resolver): ValidatedRawEvents = {
```

Comment:
nit: indent

Reply:
Fixed in https://github.com/snowplow/snowplow/issues/3532